v1/daemon/image.go

```diff
@@ -15,18 +15,64 @@
 package daemon
 
 import (
-	"fmt"
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
 
+	"github.com/google/go-containerregistry/v1/tarball"
+
+	"github.com/docker/docker/client"
 	"github.com/google/go-containerregistry/name"
 	"github.com/google/go-containerregistry/v1"
 )
 
-// TODO(dlorenc): This.
-// // image accesses an image from a remote registry
-// type image struct{}
-// var _ v1.Image = (*image)(nil)
+// image accesses an image from a docker daemon
+type image struct {
+	v1.Image
+}
+
+var _ v1.Image = (*image)(nil)
+
+// API interface for testing.
+type ImageSaver interface {
+	ImageSave(context.Context, []string) (io.ReadCloser, error)
+}
+
+// This is a variable so we can override in tests.
+var getImageSaver = func() (ImageSaver, error) {
+	return client.NewEnvClient()
+}
 
 // Image exposes an image reference from within the Docker daemon.
 func Image(ref name.Reference) (v1.Image, error) {
-	return nil, fmt.Errorf("NYI: daemon.Image(%v)", ref)
+	cli, err := getImageSaver()
+	if err != nil {
+		return nil, err
+	}
+	rc, err := cli.ImageSave(context.Background(), []string{ref.Name()})
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	imageBytes, err := ioutil.ReadAll(rc)
```

Comment:
We're storing an uncompressed Docker image in memory, ouch. :(

Could we simply make `cli.ImageSave(...)` our `Opener`?

There's clearly a speed / memory trade-off here, maybe instead we could introduce `daemon.ReadOptions{Buffer: True}` to read it into memory, and `daemon.ReadOptions{}` to reread it from the daemon for each access.

Do we expect to improve this implementation over time?  I know that we can implements parts of `v1.Image` without `docker save`, e.g. `docker inspect` gives us the `RawConfigFile()`.

To be clear: I'm totally fine with functional interfaces that are destined to be gutted and replaced, I just want to understand where we're going with this.

Reply:
the image.Inspect for RawConfigFile is a good speedup. I'll file an issue for that one.

Added a buffered mode to this version.